R/ptype.R

```diff
@@ -111,3 +111,90 @@ recipes_ptype <- function(x, ..., stage = "prep") {
   
   ptype
 }
+
+#' Validate prototype of recipe object
+#'
+#' This helper function validates a dataframe against the ptype of a recipe.
+#'
+#' @param x A `recipe` object.
+#' @param new_data A data.frame. To be patched aganist ptype of `x`.
+#' @param ... currently not used.
+#' @param stage A single character. Must be one of `"prep"` or `"bake"`. See
+#'   details for more. Defaults to `"prep"`.
+#' @param call The execution environment of a currently running function, e.g.
+#'   `caller_env()`. The function will be mentioned in error messages as the
+#'   source of the error. See the call argument of [rlang::abort()] for more
+#'   information.
+#' 
+#' @return Nothing or an error.
+#' @keywords internal
+#'
+#' @seealso [developer_functions] [recipes_ptype]
+#' 
+#' @examples
+#' rec <- recipe(mpg ~ disp, data = mtcars)
+#' 
+#' recipes_ptype_validate(rec, mtcars)
+#' 
+#' @export
+recipes_ptype_validate <- function(x,
+                                   new_data,
+                                   ...,
+                                   stage = "prep", 
+                                   call = rlang::caller_env()) {
+  old_ptype <- recipes_ptype(x, stage = stage)
+  col_names <- names(old_ptype)
+
+  new_ptype <- vctrs::vec_ptype(new_data)
+
+  if (!all(col_names %in% names(new_ptype))) {
+    offenders <- col_names[!col_names %in% names(new_ptype)]
+    cli::cli_abort(
+      "Not all variables in the recipe are present in the supplied training \\
+      set: {.var {offenders}}.",
+      call = call
+    )
+  }
+
+  new_ptype <- new_ptype[col_names]
+
+  if (!identical(lapply(old_ptype, class), lapply(new_ptype, class))) {
+    old_classes <- lapply(old_ptype, class)
+    new_classes <- lapply(new_ptype, class)
+
+    offenders <- purrr::map2_lgl(old_classes, new_classes, identical)
+    offenders <- col_names[!offenders]
+
+    msg <- c("x" = "{cli::qty(offenders)} The following variable{?s} \\
+                   has the wrong class:")
```

Comment:
```suggestion
    msg <- c("x" = "{cli::qty(offenders)} The following variable{?s} \\
                   {?has/have} the wrong class:")
```

Not sure if this order is right, but does need pluralization.